### DISCUSSION TITLE
Provide an option to inject `styles` before CSS links

### DISCUSSION THREAD

**Discussion by siriwatknp** [vercel/next.js#32663] [Ideas]
### Describe the feature you'd like to request

**The issue**

Custom styles from CSS modules are overridden by the component libraries that use CSS-in-JS approach > https://github.com/mui-org/material-ui/issues/29648#issuecomment-997190568. The desired behavior is the opposite (custom styles > component library). 

Take a look at [MUI nextjs example](https://github.com/mui-org/material-ui/blob/6af5ac9bd062fab467a7ffb3bf67f054b0d5f474/examples/nextjs-with-typescript/src/pages/_document.tsx?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L85) for the SSR setup.

I believe that this is because the styles generated by css-in-js library are injected after the `this.getCssLinks()` [here](https://github.com/vercel/next.js/blob/ce3112048ffe53b0ac8e4c292cfd8ab3bbf9c5d6/packages/next/pages/_document.tsx#L752).

It would be nice if there is a way to configure the order of the `styles`.

### Describe the solution you'd like

Have a simple flag to move styles to the top of `Head`.

### Describe alternatives you've considered

Support object type for `styles` in `getInitialProps`
- array (existing type)
- object
    - prepend: boolean // setting to `true` will inject `styles` at the top of `Head`.
    - value: the tags

```js
// _document.js

function MyDocument() {
...
}

MyDocument.getInitialProps = async () => {
  const styles = cssInJsAPIThatReturnStyleTags();
  return {
    styles: {
      prepend: true,
      value: styles,
    },
  }
}
```

**Comment by oliviertassinari** [comment]
It can be solved with a custom _document.js: https://github.com/mui-org/material-ui/pull/30381

**Reply by siriwatknp** [reply]
It does not work with App dir. The `useServerInsertedHTML()` injects the tags after links and CSS modules.
